sql/core/src/main/scala/org/apache/spark/sql/streaming/TwsTester.scala

```diff
@@ -0,0 +1,286 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.streaming
+
+import java.sql.Timestamp
+import java.time.{Clock, Instant, ZoneId}
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.sql.execution.streaming.operators.stateful.transformwithstate.statefulprocessor.ImplicitGroupingKeyTracker
+import org.apache.spark.sql.execution.streaming.operators.stateful.transformwithstate.testing.InMemoryStatefulProcessorHandle
+import org.apache.spark.sql.execution.streaming.operators.stateful.transformwithstate.timers.{ExpiredTimerInfoImpl, TimerValuesImpl}
+
+/**
+ * Testing utility for transformWithState stateful processors.
+ *
+ * This class enables unit testing of StatefulProcessor business logic by simulating the
+ * behavior of transformWithState. It processes input rows and returns output rows equivalent
+ * to those that would be produced by the processor in an actual Spark streaming query.
+ *
+ * '''Supported:'''
+ *  - Processing input rows and producing output rows via `test()`.
+ *  - Initial state setup via constructor parameter.
+ *  - Direct state manipulation via `setValueState`, `setListState`, `setMapState`.
+ *  - Direct state inspection via `peekValueState`, `peekListState`, `peekMapState`.
+ *  - Timers in ProcessingTime mode (use `advanceProcessingTime` to fire timers).
+ *  - Timers in EventTime mode (use `eventTimeExtractor` and `watermarkDelayMs` to configure;
+ *    watermark advances automatically based on event times, or use `advanceWatermark` manually).
```

Comment:
> watermark advances automatically based on event times, or use `advanceWatermark` manually

Both conditions taking effect to watermark lead confusion. Why not just picking the latter one? Is the former coupled with the implementation?

Reply:
It's not just about confusion. Since we only support test() with a single key, it's not going to test the case where the watermark is advanced "after" multiple grouping keys have been processed, which is the default behavior of microbatch mode.

I'd rather leave users to advance watermark for both 1) consistency with processing time timer and 2) closer to actual behavior.